frame/support/procedural/src/pallet_version.rs

```diff
@@ -19,46 +19,59 @@
 
 use proc_macro2::{TokenStream, Span};
 use syn::{Result, Error};
-use std::{env, str::FromStr};
+use std::{convert::TryInto, env, fmt::Display, path::PathBuf};
 use frame_support_procedural_tools::generate_crate_access_2018;
-
-/// Get the version from the given version environment variable.
-///
-/// The version is parsed into the requested destination type.
-fn get_version<T: FromStr>(version_env: &str) -> std::result::Result<T, ()> {
-	let version = env::var(version_env)
-		.unwrap_or_else(|_| panic!("`{}` is always set by cargo; qed", version_env));
-
-	T::from_str(&version).map_err(drop)
-}
+use cargo_metadata::MetadataCommand;
 
 /// Create an error that will be shown by rustc at the call site of the macro.
-fn create_error(message: &str) -> Error {
+fn create_error(message: impl Display) -> Error {
 	Error::new(Span::call_site(), message)
 }
 
-/// Implementation of the `crate_to_pallet_version!` macro.
-pub fn crate_to_pallet_version(input: proc_macro::TokenStream) -> Result<TokenStream> {
+/// Implementation of the `pallet_version!` macro.
+pub fn pallet_version(input: proc_macro::TokenStream) -> Result<TokenStream> {
 	if !input.is_empty() {
 		return Err(create_error("No arguments expected!"))
 	}
 
-	let major_version = get_version::<u16>("CARGO_PKG_VERSION_MAJOR")
-		.map_err(|_| create_error("Major version needs to fit into `u16`"))?;
+	let cargo_toml = PathBuf::from(
+		env::var("CARGO_MANIFEST_DIR").expect("`CARGO_MANIFEST_DIR` is set by cargo"),
+	).join("Cargo.toml");
 
-	let minor_version = get_version::<u8>("CARGO_PKG_VERSION_MINOR")
-		.map_err(|_| create_error("Minor version needs to fit into `u8`"))?;
+	let metadata = MetadataCommand::new()
+		.manifest_path(cargo_toml)
+		.exec()
+		.expect("`cargo metadata` can not fail on project `Cargo.toml`; qed");
 
-	let patch_version = get_version::<u8>("CARGO_PKG_VERSION_PATCH")
-		.map_err(|_| create_error("Patch version needs to fit into `u8`"))?;
+	let package = metadata.root_package().expect("We specified a manifest path; qed");
 
-	let crate_ = generate_crate_access_2018("frame-support")?;
+	let version: u16 = if let Some(frame) = package.metadata.get("frame").and_then(|f| f.as_object()) {
+		if let Some(key) = frame.keys().find(|k| *k != "pallet-version") {
+			return Err(create_error(format!("Unknown key in package.metadata.frame: {}", key)))
+		}
 
-	Ok(quote::quote! {
-		#crate_::traits::PalletVersion {
-			major: #major_version,
-			minor: #minor_version,
-			patch: #patch_version,
+		if let Some(pallet_version) = frame.get("pallet-version") {
+			if let Some(pallet_version) = pallet_version.as_u64() {
+				pallet_version
+					.try_into()
+					.map_err(|_|
+						create_error(
+							"package.metadata.frame.pallet-version supports in maximum u16."
+						)
+					)?
+			} else {
+				return Err(
+					create_error("package.metadata.frame.pallet-version is required to be a number"),
+				)
+			}
+		} else {
+			1
```

Comment:
This value is supposed to be stored in storage. My default would be that this macro would similarly return `0` if none is defined.

Reply:
Agreed that the default value should be 0.